l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,720 @@
+            // Determine if the given dataHash matches the calculated legacy transaction hash.
+            bool isLegacyTxDataHash = _isLegacyTxDataHash(_depositSender, _assetId, _assetData, dataHash);
+            // If the dataHash matches the legacy transaction hash, skip the next step.
+            // Otherwise, perform the check using the new transaction data hash encoding.
+            if (!isLegacyTxDataHash) {
+                bytes32 txDataHash = DataEncoding.encodeTxDataHash({
+                    _encodingVersion: NEW_ENCODING_VERSION,
+                    _prevMsgSender: _depositSender,
+                    _assetId: _assetId,
+                    _nativeTokenVault: address(l1NativeTokenVault),
+                    _transferData: _assetData
+                });
+                if (dataHash != txDataHash) {
+                    revert DepositDoesNotExist();
+                }
+            }
+        }
+        delete depositHappened[_chainId][_l2TxHash];
+    }
+
+    /// @notice Finalize the withdrawal and release funds.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    /// @dev We have both the legacy finalizeWithdrawal and the new finalizeDeposit functions,
+    /// finalizeDeposit uses the new format. On the L2 we have finalizeDeposit with new and old formats both.
+    function finalizeDeposit(FinalizeL1DepositParams calldata _finalizeWithdrawalParams) public {
+        _finalizeDeposit(_finalizeWithdrawalParams);
+    }
+
+    /// @notice Internal function that handles the logic for finalizing withdrawals, supporting both the current bridge system and the legacy ERC20 bridge.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    function _finalizeDeposit(
+        FinalizeL1DepositParams calldata _finalizeWithdrawalParams
+    ) internal nonReentrant whenNotPaused {
+        bytes memory transferData;
+        bytes32 assetId;
+        (assetId, transferData) = _verifyAndGetWithdrawalData(_finalizeWithdrawalParams);
+        l1AssetRouter.finalizeDeposit(_finalizeWithdrawalParams.chainId, assetId, transferData);
+    }
+
+    /// @notice Internal function that handles the logic for finalizing withdrawals, supporting both the current bridge system and the legacy ERC20 bridge.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    /// @return assetId The bridged asset ID.
+    /// @return transferData The encoded transfer data.
+    function _verifyAndGetWithdrawalData(
+        FinalizeL1DepositParams calldata _finalizeWithdrawalParams
+    ) internal whenNotPaused returns (bytes32 assetId, bytes memory transferData) {
+        if (
+            isWithdrawalFinalized[_finalizeWithdrawalParams.chainId][_finalizeWithdrawalParams.l2BatchNumber][
+                _finalizeWithdrawalParams.l2MessageIndex
+            ]
```

Comment:
1. It is a bit hard to read because of big naming
2. You can save a bit of gas and improve naming by caching the parameters into stack variables:

```
uint256 chainId = _finalizeWithdrawalParams.chainId;
...
```